telegram/ext/_applicationbuilder.py

```diff
@@ -0,0 +1,766 @@
+        self._arbitrary_callback_data = arbitrary_callback_data
+        return self
+
+    def bot(
+        self: 'ApplicationBuilder[BT, CCT, UD, CD, BD, JQ]',
+        bot: InBT,
+    ) -> 'ApplicationBuilder[InBT, CCT, UD, CD, BD, JQ]':
+        """Sets a :class:`telegram.Bot` instance to be used for
+        :attr:`telegram.ext.Application.bot`. Instances of subclasses like
+        :class:`telegram.ext.ExtBot` are also valid.
+
+        Args:
+            bot (:class:`telegram.Bot`): The bot.
+
+        Returns:
+            :class:`ApplicationBuilder`: The same builder with the updated argument.
+        """
+        if self._updater not in (DEFAULT_NONE, None):
+            raise RuntimeError(_TWO_ARGS_REQ.format('bot', 'updater'))
+        for attr, error in _BOT_CHECKS:
+            if not isinstance(getattr(self, f'_{attr}'), DefaultValue):
+                raise RuntimeError(_TWO_ARGS_REQ.format('bot', error))
+        self._bot = bot
+        return self  # type: ignore[return-value]
+
+    def update_queue(self: BuilderType, update_queue: Queue) -> BuilderType:
+        """Sets a :class:`queue.Queue` instance to be used for
+        :attr:`telegram.ext.Application.update_queue`, i.e. the queue that the application will
+        fetch updates from. Will also be used for the :attr:`telegram.ext.Application.updater`.
+        If not called, a queue will be instantiated.
+
+         .. seealso:: :attr:`telegram.ext.Updater.update_queue`
+
+        Args:
+            update_queue (:class:`queue.Queue`): The queue.
+
+        Returns:
+            :class:`ApplicationBuilder`: The same builder with the updated argument.
+        """
+        if self._updater not in (DEFAULT_NONE, None):
+            raise RuntimeError(_TWO_ARGS_REQ.format('update_queue', 'updater instance'))
+        self._update_queue = update_queue
+        return self
+
+    def concurrent_updates(self: BuilderType, concurrent_updates: Union[bool, int]) -> BuilderType:
+        """Specifies if and how many updates may be processed concurrently instead of one by one.
+
+        Warning:
+            Processing updates concurrently is not recommended when stateful handlers like
+            :class:`telegram.ext.ConversationHandler` are used.
+
+         .. seealso:: :paramref:`telegram.ext.Application.concurrent_updates`
+
+        Args:
+            concurrent_updates (:obj:`bool` | :obj:`int`): Passing :obj:`True` will allow for 4096
+                updates to be processed concurrently. Pass an integer to specify a different number
+                of updates that may be processed concurrently.
+
+        Returns:
+            :class:`ApplicationBuilder`: The same builder with the updated argument.
+        """
+        self._concurrent_updates = concurrent_updates
+        return self
+
+    def job_queue(
+        self: 'ApplicationBuilder[BT, CCT, UD, CD, BD, JQ]',
+        job_queue: InJQ,
+    ) -> 'ApplicationBuilder[BT, CCT, UD, CD, BD, InJQ]':
```

Comment:
iisc, `JQ` and `InJQ` are the same? Also why not just use `BuilderType` as was used above

Reply:
`In` = `Input` … Naming is hard xD Problem is that mypy needs to know that `self` currently has a different type of jq than what the type of `job_queue` is, so using the same typevar for both doesn't work. This also explains why using `BuilderType` doesn't do the trick: We want to specify that one of the generic slots of `BuilderType` changes the type :)